HEN_HOUSE/egs++/geometry/egs_cd_geometry/egs_cd_geometry.h

```diff
@@ -506,9 +506,25 @@ class EGS_CDGEOMETRY_EXPORT EGS_CDGeometry : public EGS_BaseGeometry {
                     }
                     return ixold;
                 }
-                else if (ixnew < 0 && tb <= epsilon) {                         // (a) is true
+				// Skip for concave geometries: particles can reenter the CD geometry after leaving.
+				// Check 2. below will catch these cases.
+                else if ((ixnew < 0 && tb <= epsilon) && (bg->isConvex() && g[ibase]->isConvex())) {                         // (a) is true
                     return ixnew;
```

Comment:
I'm not entirely sure if those checks will cover everything... Another solution which seemed to work was to move the check for (a) to below check 2. This allows 2. to properly catch the concave geometry cases and then fall back on the check for condition (a), if needed. Thoughts?